draft-ietf-mls-protocol.md

```diff
@@ -948,6 +948,212 @@ update_secret -> HKDF-Extract = epoch_secret
+    raw bytes of MLSPlaintext.application_data
+  * signature: The raw bytes of MLSPlaintext.signature
+  * sig_len: The length of MLSPlaintext.signature
+  * type: The value of MLSPlaintext.type
+  * zero_padding: An all-zero octet string of any length dictated by
+    the application (including zero length)
+3. Construct a new MLSCiphertext object with the following contents:
+  * group_id, epoch, sender: The corresponding values from the
+    MLSPlaintext
+  * generation: The current generation in the application key
+    schedule
+  * ciphertext: The encryption of the MLSInnerPlaintext using the
+    AEAD cipher for the group's ciphersuite, and the key and nonce
+    specified by the group_id, epoch, sender, and generation
+
+To decrypt an MLSCiphertext object, the sender takes the following
+steps:
+
+1. Use the group_id, epoch, sender, and generation to look up the
+   appropritate key and nonce.
+2. Attempt to decrypt MLSCiphertext.ciphertext using the AEAD cipher
+   for the group's ciphersuite.  If decryption fails, abort this
+   process and return an error.
+3. Parse the decrypted value as an MLSInnerPlaintext value, starting
+   from the end of the decrypted value:
+   * zero_padding: The continuous string of zero octets at the end
+     of the decrypted value
+   * type: The value of the octet immediately preceding the
+     zero_padding field
+   * sig_len: The two octets preceding the type
+   * signature: The sig_len octets preceding the sig_len field
+   * content: The remaining octets preciding the signature
+4. If an invalid content type or signature length was found in the
+   previous step, abort this process and return an error
+5. Construct a new MLSPlaintext value with the following contents:
+   * group_id, epoch, sender: The corresponding values from the
+     MLSCiphertext
+   * type: The value of MLSInnerPlaintext.type
+   * handshake or application: Populated from
+     MLSInnerPlaintext.content according to the type
+   * signature: The value of MLSInnerPlaintext.signature
+6. Use the sender field to look up the appropriate Credential in the
+   roster, and use it to verify the signature on the MLSPlaintext.
+   If verifiation fails, abort this process and return an error.
+
+The following figure illustrates the flow of inforormation between
+MLSCiphertext, MLSInnerPlaintext, and MLSPlaintext values:
+
+~~~~~
+MLSPlaintext                            MLSCiphertext
+  group_id  |                           | group_id
+  epoch     |---------------------------| epoch
+  sender    |                           | sender
+
+                                        * generation
+  type      |
+  content   |--+                     +--> ciphertext
+  signature |  |  MLSInnerPlaintext  |
+               |  | content       |  |
+               +--| signature     |--+
+                  | sig_len       |
+                  | type          |
+                                  |
+                  * zero_padding  |
+~~~~~
```

Comment:
Do I understand correctly that the fields that represent information that is shared between the different structs are not actually part of the encoded data structure but rather are just meant to be pointers to the respective "outer" struct? If that is the case, I think we should stay true to the principle of having the structs represent what is actually part of the data structure that is encoded and not what is composed locally to compute signatures. Please do correct me if I'm wrong in my understanding of the spec here.